---
title: "Genomic data in R"
author: "LinnÃ©a Smeds"
date: "`r format(Sys.Date(),format='%d/%m/%Y')`"
output: ioslides_presentation
logo: "images/raukR_logo.svg"

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE)
knitr::opts_chunk$set(cache=TRUE)
```

# Part 1  | Data and Aims

## Data

### Whole genome SNP data from 210 wolves:
    - *Scandinavian=Recently founded*
    - *Finnish/Russian=Source population*
   
### Detailed information about Scandinavian population: 
  - *sample year*
  - *founding individuals*
  - *generation to founder*
  
> (in this project, I only use ~5000 deleterious SNPs)
  
## Aim

To rewrite bash&python code that extract frequencies of homozygous derived/heterozygous sites for each individual from the vcf file, and plot population differences along with all info given for the focal population

### My personal goal

Learn how to use tidyverse (especially pipes) and ggplot

# Part 2 | Results

## Packages used
```{r, message=FALSE}
require(vcfR)
require(tictoc)
require(viridis)
require(tidyverse)
```

## Read in the vcf file and convert it to tidy 
```{r, message=FALSE, eval=FALSE}
vcf_file="data/Wolf.deleterious.missense.polarized.vcf"
vcf <- vcfR::read.vcfR(vcf_file)
tidy_vcf <- vcfR::vcfR2tidy(vcf, 
                      info_fields=c("AA"), 
                      format_fields=c("GT"), 
                      dot_is_NA=TRUE)

```
The population, founder and generation info from plain textfiles are saved as tibbles directly
```{r, eval=FALSE}
# example
pop_tib<- "data/populations.txt" %>% 
      read.table() %>% 
      as_tibble() %>% 
      rename(Indiv=V1, Pop=V2)
```


```{r, include=FALSE}
vcf_file="data/Wolf.deleterious.missense.polarized.vcf"
vcf <- vcfR::read.vcfR(vcf_file)
tidy_vcf <- vcfR::vcfR2tidy(vcf, 
                      info_fields=c("AA"), 
                      format_fields=c("GT"), 
                      dot_is_NA=TRUE)

```
## Inspecting the tables
```{r}
str(tidy_vcf)
```

## Inspecting the tables (cont.)
```{r}
head(tidy_vcf$gt, 3)
head(tidy_vcf$fix, 3)
```

```{r, eval=TRUE, include=FALSE}
pop_file="data/5TempClasses11I95F15R.pop.list"
founder_file="data/founders.list"
gen_file="data/generation.list"
pop_tib<-pop_file %>% read.table() %>% as_tibble() %>% rename(Indiv=V1, Pop=V2)
gen_tib<-gen_file %>% read.table(header=TRUE) %>% as_tibble() %>% rename(Generation=generation)
fou_tib <- founder_file %>% read.table(header=TRUE) %>% as_tibble() %>% rename(Founder=Ftype)
```

## Extract relevant info - Trial #1
```{r, message=FALSE}
tic()
okpos<- tidy_vcf$fix %>%filter(REF==AA)
invertpos<-tidy_vcf$fix %>%filter(ALT==AA)
homref_aa_ref<-tidy_vcf$gt %>% inner_join(okpos) %>% filter(gt_GT=="0/0") %>% group_by(Indiv) %>%summarize(num_homref_aa=n())
homalt_aa_ref<-tidy_vcf$gt %>% inner_join(okpos) %>% filter(gt_GT=="1/1") %>% group_by(Indiv) %>%summarize(num_homalt_der=n())
homref_aa_alt<-tidy_vcf$gt %>% inner_join(invertpos) %>% filter(gt_GT=="0/0") %>% group_by(Indiv) %>%summarize(num_homref_der=n())
homalt_aa_alt<-tidy_vcf$gt %>% inner_join(invertpos) %>% filter(gt_GT=="1/1") %>% group_by(Indiv) %>%summarize(num_homalt_aa=n())
aa_tib<- homref_aa_ref %>% inner_join(homalt_aa_alt) %>% mutate(num_aa=num_homref_aa+num_homalt_aa) %>% select(Indiv, num_aa)
der_tib<-homalt_aa_ref %>% inner_join(homref_aa_alt) %>% mutate(num_der=num_homref_der+num_homalt_der) %>% select(Indiv, num_der)
het<-tidy_vcf$gt %>% filter(gt_GT=="0/1") %>% group_by(Indiv) %>%summarize(num_het=n())
tidy_gt<-aa_tib %>% inner_join(het) %>% inner_join(der_tib) 
toc()
```

## Extract relevant info - Trial #2
```{r, message=FALSE}
tic()
refAA_tib<-tidy_vcf$gt %>% filter(!is.na(gt_GT)) %>% inner_join(okpos) %>% group_by(Indiv, gt_GT) %>% summarize(count=n()) %>% ungroup() %>% mutate(gt = case_when(gt_GT=='0/0' ~ 'refAA_homanc', gt_GT=='0/1' ~'refAA_het', gt_GT=='1/1' ~'refAA_homder')) %>% select(-gt_GT) %>% pivot_wider(names_from=gt, values_from=count)
altAA_tib<-tidy_vcf$gt %>% filter(!is.na(gt_GT)) %>% inner_join(invertpos) %>% group_by(Indiv, gt_GT) %>% summarize(count=n()) %>% ungroup() %>% mutate(gt = case_when(gt_GT=='0/0' ~ 'altAA_homder', gt_GT=='0/1' ~'altAA_het', gt_GT=='1/1' ~'altAA_homanc')) %>% select(-gt_GT) %>% pivot_wider(names_from=gt, values_from=count)
tidy_gt2 <- refAA_tib %>% inner_join(altAA_tib) %>% transmute(Indiv, num_aa=altAA_homanc+refAA_homanc, num_het=refAA_het+altAA_het, num_der=refAA_homder+altAA_homder) 
toc()
```

## Extract relevant info - Trial #3
```{r, message=FALSE}
tic()
tidy_gt3<-tidy_vcf$gt %>% filter(!is.na(gt_GT)) %>% 
  inner_join(tidy_vcf$fix)  %>% 
  mutate(new_gt=case_when((ALT==AA & gt_GT=='0/0') ~'1/1',
                          (ALT==AA & gt_GT=='1/1') ~'0/0',
                          TRUE ~ gt_GT)) %>%
  group_by(Indiv, new_gt) %>% summarize(count=n()) %>% 
  ungroup() %>% mutate_if(is.character, str_replace_all, 
                          pattern = "/", replacement = "") %>%
  pivot_wider(names_from=new_gt, names_prefix="gt", values_from=count)
toc()
```

## Finally!
```{r}
head(tidy_gt3, 3)
```
Merge with other info tibbles
```{r, message=FALSE}
final_gt<- tidy_gt3 %>% mutate(sum=gt00+gt01+gt11) %>% 
  inner_join(pop_tib) %>% 
  mutate(hetfrq=gt01/sum, homderfrq=gt11/sum, 
         deralfreq=(gt11*2+gt01)/(sum*2),
         Grp=case_when(Pop=='Finland' | Pop=='Russia' ~ Pop, TRUE ~ 'Scandinavia'))  %>% 
  left_join(fou_tib) %>% left_join(gen_tib)
```

## Now to the plotting - boxes and points
```{r}
ggplot(final_gt,aes(Pop,hetfrq)) + geom_boxplot(aes(fill=Pop)) +
    geom_point() +labs(x=NULL,y="proportion heterozygous sites")+
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())
```

## Reorder and add facets
```{r}
final_gt$Pop <- factor(final_gt$Pop, levels = c("1983-1990", "1991-1998", "1999-2006", "2007-2014S", "2007-2014I", "Immigrant", "Finland", "Russia"))
ggplot(final_gt,aes(Pop, hetfrq))+geom_boxplot(aes(fill=Pop))+
  geom_point() + facet_grid(~factor(Grp, level=c("Scandinavia", "Finland", "Russia")), scales="free_x", space="free_x") + labs(x=NULL,y="proportion heterozygous sites")+
  theme(legend.title=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
```

## Violin, changing colours etc
```{r}
ggplot(final_gt,aes(Pop, hetfrq))+  geom_violin(aes(fill=Pop))+  geom_point() + scale_fill_viridis(discrete=TRUE) + facet_grid(~factor(Grp, level=c("Scandinavia", "Finland", "Russia")), scales="free_x", space="free_x") +  labs(x=NULL,y="proportion heterozygous sites")+theme(panel.grid.major = element_line(colour = 'white'),      panel.background = element_rect(fill = '#f5f4e6', colour = '#FFFDF8'),legend.title=element_blank(), axis.text.x=element_blank(),axis.ticks.x=element_blank())
```

## Adding founder and generation info
```{r}
p<- ggplot(final_gt,aes(Pop, hetfrq))+geom_violin(aes(fill=Pop))+ scale_fill_viridis(discrete=TRUE) +facet_grid(~factor(Grp, level=c("Scandinavia", "Finland", "Russia")), scales="free_x", space="free_x") + labs(x=NULL,y="proportion heterozygous sites")+ theme(panel.grid.major = element_line(colour = 'white'),panel.background = element_rect(fill = '#f5f4e6', colour = '#FFFDF8'),axis.text.x=element_blank(),axis.ticks.x=element_blank(),legend.box='vertical')
p+geom_point(data=filter(final_gt, !is.na(Founder)),aes(shape=Founder), size=3, col='darkgrey', fill='lightyellow')+  scale_shape_manual(values=c(22,24))+theme(legend.key=element_rect(fill='#3acf73', colour='NA')) + geom_point(data=filter(final_gt, !is.na(Generation)), aes(color=Generation)) +   scale_colour_continuous(low='#edf9ff', high='#008ed9', breaks = c(1,2,3,4,5,6), labels = c(1,2,3,4,5,6))
```

# Conclusions

## It worked!

> - Converted 95 lines of python to 4 lines of R code

> - Learned how to make (very long) pipes

> - Started to dig into the world of ggplots

# Thank you
<img src="images/female_wolf_Ilpo Kojola.jpg" width="350px">
